### PR TITLE
Changes required for 4.1.0-rc3

### PR DESCRIPTION
**Pull request by stbenjam**
Changes required for update to RC.3

  - [ ] This PR
  - [x] https://github.com/openshift-metal3/kni-installer/pull/77
  - [x] Libvirt api-int record: https://github.com/openshift-metal3/dev-scripts/pull/536, https://github.com/metal3-io/metal3-dev-env/pull/9
  - [ ] Handle api-int case for baremetal: TBD
  - [ ] https://github.com/openshift-metal3/kni-installer/pull/75
  - [x] Resolve machine-api failure https://github.com/openshift-metal3/dev-scripts/issues/537
  - [ ] Change name of machine labels https://github.com/openshift-metal3/kni-installer/pull/80

### CODE REVIEW

**Comment by russellb** [comment]
looks like it should be on quay.io/openshift-release-dev/ocp-release

**Comment by stbenjam** [comment]
It does work with this url, but seems to end up redirecting to quay in the end

**Comment by intlabs** [comment]
> Switching will drop the need for the extra CI registry pull secret

Confirmed:
$ docker pull quay.io/openshift-release-dev/ocp-release:4.1.0-rc.1
4.1.0-rc.1: Pulling from openshift-release-dev/ocp-release
5dfdb3f0bcc0: Pull complete
99f178453a43: Pull complete
f9fa6d5c56a6: Pull complete
d252d93af3e8: Pull complete
a9607047a865: Pull complete
Digest: sha256:be61a9ec132118e41a417b347242361d9cc96b1a73753e121dc7a74a1905baea
Status: Downloaded newer image for quay.io/openshift-release-dev/ocp-release:4.1.0-rc.1

**Comment by russellb** [comment]
We should also look at https://github.com/openshift-metal3/dev-scripts/pull/401 - if that's ready, would be good to base our next update on this new approach

**Comment by stbenjam** [comment]
I’ve been watching the work on that. I agree, would like to use that process. Looks like a few things left to do on the checklist at the top of the PR before it’s ready.

File: assets/templates/99_registry.yaml [hunk 1]
```diff
@@ -15,6 +15,4 @@ spec:
     write:
       maxWaitInQueue: 0s
   storage:
-    filesystem:
-      volumeSource:
-        emptyDir: {}
+    emptyDir: {}
```
Comment (stbenjam) [inline assets/templates/99_registry.yaml]: This changed: https://github.com/openshift/cluster-image-registry-operator/commit/cc4403d6e

File: vm-setup/roles/common/tasks/main.yml [hunk 2]
```diff
@@ -68,6 +68,7 @@
                 - ip: "{{ baremetal_network_cidr | nthhost(5) }}"
                   hostnames:
                     - "api"
+                    - "api-int"
```
Comment (stbenjam) [inline vm-setup/roles/common/tasks/main.yml]: This change is still needed, at least for now - the masters are pointing at dnsmasq when they come up and they get their ignition config from api-int

File: utils.sh [hunk 3]
```diff
@@ -211,7 +211,7 @@ function patch_ep_host_etcd() {
     declare -r domain="$1"
     declare -r srv_record="_etcd-server-ssl._tcp.$domain"
     declare -r api_domain="api.$domain"
-    host_etcd_ep=$(oc get ep -n kube-system host-etcd -o json | jq -r '{"subsets": .subsets}')
+    host_etcd_ep=$(oc get ep -n openshift-etcd host-etcd -o json | jq -r '{"subsets": .subsets}')
```
Comment (stbenjam) [inline utils.sh]:
```console
[root@dell-r730-021 dev-scripts]# oc get ep -n kube-system host-etcd -o json | jq -r '{"subsets": .subsets}'
Error from server (NotFound): endpoints "host-etcd" not found

[root@dell-r730-021 dev-scripts]# oc get ep -n openshift-etcd host-etcd -o json | jq -r '{"subsets": .subsets}'
{
  "subsets": [
    {
      "addresses": [
        {
          "hostname": "etcd-0",
          "ip": "192.168.111.20"
        },
        {
          "hostname": "etcd-1",
          "ip": "192.168.111.21"
        },
        {
          "hostname": "etcd-2",
          "ip": "192.168.111.22"
        }
      ],
      "ports": [
        {
          "name": "etcd",
          "port": 2379,
          "protocol": "TCP"
        }
      ]
    }
  ]
}
```

**Comment by stbenjam** [comment]
I was able to get a successful 4.1.0 rc.3 install, using the PR's
listed below.  I didn't do any deep testing but my cluster looks
functional.

  - https://github.com/openshift-metal3/dev-scripts/pull/523
  - https://github.com/openshift-metal3/kni-installer/pull/77
  - https://github.com/openshift-metal3/kni-installer/pull/75

File: vm-setup/roles/common/tasks/main.yml (hunk 2 above)
Comment (hardys) [reply vm-setup/roles/common/tasks/main.yml]: I guess this is fine as an interim solution in the  VM case, but how do we avoid requiring a new DNS record for the API VIP in the baremetal case?  cc @celebdor and @cybertron

Comment (celebdor) [reply vm-setup/roles/common/tasks/main.yml]: The solution is to get the VIPs from install config and serve them in coredns (and make ignition fetch from VIP instead of api)

Comment (cybertron) [reply vm-setup/roles/common/tasks/main.yml]: https://github.com/openshift-metal3/kni-installer/pull/77 makes api-int available from coredns. We'll still need the change to point ignition at the VIP instead of a DNS name though.

Comment (stbenjam) [reply vm-setup/roles/common/tasks/main.yml]: If I'm looking in the right place, doesn't ignition happen using HTTPS, verifying the hostname and certificate? The endpoint is also not platform specific: how do we make that work?

https://github.com/openshift/installer/blob/master/pkg/asset/ignition/machine/node.go#L23-L25

**Comment by hardys** [comment]
Ah sorry this will need a rebase since vm-setup moved to metal3-dev-env - I wonder if we should parameterize some of `vm-setup/roles/common/tasks/main.yml ` so that we can control the DNS things which are specific to OpenShift via the `vm_setup_vars.yml` in dev-scripts?

File: ocp_install_env.sh [hunk 4]
```diff
@@ -13,7 +13,7 @@ export EXTERNAL_SUBNET="192.168.111.0/24"
 # The release we default to here is pinned and known to work with our current
 # version of kni-installer.
 #
-export OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="registry.svc.ci.openshift.org/ocp/release:4.0.0-0.ci-2019-04-17-133604"
+export OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE="registry.svc.ci.openshift.org/ocp/release:4.1.0-rc.3"
```
Comment (markmc) [inline ocp_install_env.sh]: Looks like from the conversation that we agree to use the URL listed on e.g. https://openshift-release.svc.ci.openshift.org/releasestream/4-stable/release/4.1.0-rc.1 - i.e. `quay.io/openshift-release-dev/ocp-release:4.1.0-rc.1`

**Comment by markmc** [comment]
@stbenjam want to add a checklist at the top listing what needs to be done before we can merge? Finding it hard to follow ...

File: assets/templates/99_registry.yaml (hunk 1 above)
Comment (markmc) [reply assets/templates/99_registry.yaml]: Let's get this useful context in the commit message - it'll get lost here after the PR is merged

File: utils.sh (hunk 3 above)
Comment (markmc) [reply utils.sh]: openshift/installer#1665 is a good reference on the background for this change

**Comment by hardys** [comment]
> @stbenjam want to add a checklist at the top listing what needs to be done before we can merge? Finding it hard to follow ...

Sounds good, but I think the main remaining item is to resolve the api-int DNS solution, what's proposed here will only work for VMs AFAICS and we want to avoid requiring those testing baremetal to create another upstream DNS record for the "internal" VIP.

So we need to figure out how ignition can resolve the IP in both VM and baremetal cases, which as mentioned by @stbenjam is tricky as I don't think we can just replace the hostname with an IP, and the resolv.conf additions don't happen until after the ignition/pivot happens, so we have a chicken/egg problem?

**Comment by stbenjam** [comment]
I updated the description with the checklist.  I'll rebase this PR when I'm at my workstation.  Essentially, 
 nothing's changed since https://github.com/openshift-metal3/dev-scripts/pull/523#issuecomment-491968815 - but there are a number of open and WIP PR's related to the `api-int` change. It's not clear to me if @cybertron @celebdor want to get something else in, or if they are ok with only the 3 listed above which worked for me.

**Comment by markmc** [comment]
> > @stbenjam want to add a checklist at the top listing what needs to be done before we can merge? Finding it hard to follow ...
> 
> Sounds good, but I think the main remaining item is to resolve the api-int DNS solution, what's proposed here will only work for VMs AFAICS and we want to avoid requiring those testing baremetal to create another upstream DNS record for the "internal" VIP.
> 
> So we need to figure out how ignition can resolve the IP in both VM and baremetal cases, which as mentioned by @stbenjam is tricky as I don't think we can just replace the hostname with an IP, and the resolv.conf additions don't happen until after the ignition/pivot happens, so we have a chicken/egg problem?

Thanks for the explanation.

I suggest we proceed with rebasing ASAP and add "remove the need for an api-int DNS entry on baremetal" to our backlog?

**Comment by stbenjam** [comment]
@hardys I rebased, and moved the api-int change here: https://github.com/metal3-io/metal3-dev-env/pull/8

That's a bit specific to openshift, isn't it? Maybe we should parametrize those records in that role?

**Comment by hardys** [comment]
> @hardys I rebased, and moved the api-int change here: [metal3-io/metal3-dev-env#8](https://github.com/metal3-io/metal3-dev-env/pull/8)
> 
> That's a bit specific to openshift, isn't it? Maybe we should parametrize those records in that role?

Thanks! And yes, as mentioned in my earlier comment we probably should parameterize/sanitize some of that but I missed it in my earlier refactoring - I can take a look unless you've already started something?

File: assets/templates/99_registry.yaml (hunk 1 above)
Comment (stbenjam) [reply assets/templates/99_registry.yaml]: Done

**Comment by hardys** [comment]
I just pushed https://github.com/openshift-metal3/dev-scripts/pull/536 which would enable us to add api-int via a variable only to dev-scripts, it will need to be merged together with https://github.com/metal3-io/metal3-dev-env/pull/9

File: vm-setup/roles/common/tasks/main.yml (hunk 2 above)
Comment (cybertron) [reply vm-setup/roles/common/tasks/main.yml]: We also need to tweak the certificate so it includes the VIP in addition to the DNS name. Some work has been done in that direction by other teams and I need to look at getting it integrated into our fork of the installer.

**Comment by markmc** [comment]
We should be able to proceed with registry.svc.ci.openshift.org/kni/release:4.1.0-rc.3-kni.0 as a temporary workaround for #537 - see https://gist.github.com/markmc/f8a78d7cea7252a9e0f29dadcfaa1253 for how it was built

**Comment by stbenjam** [comment]
Updated the pin, and I'm testing a new deployment right now.  Thanks!

**Comment by russellb** [comment]
> It comes up fine, but hitting an issue on worker deployments. Based on discussion in slack, it sounds like it's pre-existing on 4.0 as well. The worker node shows "Node not found hook failed: Port 00:f6:ff:4d:bb:31 already exists" when it boots

This isn't the real error.  You see this because we haven't actually provisioned the worker host at all, and the baremetal-operator powers on hosts that aren't provisioned if we have `online: true` set on them.  It's questionable whether that behavior makes sense, but it's behaving as expected.

The real error is up in the cluster-api integration that prevented the worker from being provisioned.  See https://github.com/openshift-metal3/dev-scripts/issues/537#issuecomment-493116626
